### ISSUE TITLE
Topics with processing-role="resource-only" generate XHTML

### ISSUE THREAD

**Issue by robander** [dita-ot/dita-ot#2077]
Reported to me by a user (not sure which OT version), but I replicated with 2.1.1. If a map has nothing but key definitions (that is, nothing but references with `processing-role="resource-only"`), XHTML output still generates result files for each of the topics. Resource-only topics should be available to the transform for whatever reason, but should not be generating output.

To reproduce: in `hierarchy.ditamap`, add `processing-role="resource-only"` to every `<topicref>`. Build to XHTML; all topics are still generated. PDF output fails, I assume because no real content goes in the PDF. I tested this test both with the original `<topicref>` (add `@processing-role` manually) and with `<keydef>` (pick up default `@processing-role` and add `@keys`).

If I set the first topic in the map to `@processing-role="normal"`, but leave the others as `resource-only`, XHTML still generates all topics. PDF correctly includes only the one normal topic.

Sample set: 
[2077.zip](https://github.com/dita-ot/dita-ot/files/1005790/2077.zip)

**Comment by eerohele** [comment]
#1734 sounds related.

**Comment by robander** [comment]
Yes, you're right - it is. (I did a search on `resource-only` but that topic didn't sound like the same thing from the title.)

The other one adds in the complicating factor of `@conref` - I've always found it strange that if you just use `@conref` to refer to a topic, that other topic generates X/HTML. But, that's been the case since day 1, and the `onlytopic.in.map` parameter was added to get around it.

**Comment by robander** [comment]
I looked into this more with 2.2.2 after another user reported similar issues. I don't think my original description was correct. 

If I use the sample map and set every topic to `resource-only`, only *some* of the topics are generated. I haven't verified with every topic and link, but it looks like if a `resource-only` topic links to another topic, that other topic is generated. So, when I have these:
```xml
<topicref processing-role="resource-only" href="tasks/changingtheoil.xml" type="task"/>
<topicref processing-role="resource-only" href="concepts/oil.xml" type="concept"/>
```
When I build to HTML5 or XHTML, the output file `concepts/oil.html` is generated because there is a link to it in `tasks/changingtheoil.xml`. The reverse is also true, a link from `oil.xml` results in the output file `tasks/changingtheoil.html`

So, if a resource-only file links to another DITA topic, that other topic is generated. There does not appear to be a way to turn this off. Setting the other file to resource-only does not work, and the `onlytopic.in.map` parameter is not working. This also makes it look even more like the issue reported in #1734 where a file only used for conref also has a link, and that link target is always generated.

**Comment by Tr1st4n** [comment]
:+1: I encounter the same issue when creating webhelp output.

There is output created for topic B from which an element is 'conrefed' by an element in topic A.
Topic A is attached via a keydef and is nowhere else linked in the ditamap.
For Topic A no output created.

Note the same is applicable if topic B is linked to by an XREF in topic A.
switching onlytopic.in.map on or of makes no difference.

More information on DITA Users List:
https://groups.yahoo.com/neo/groups/dita-users/conversations/messages/40500

**Comment by robander** [comment]
Just ran into another variant of this. Very simple trimmed down use case:

- I have a map with one topic `original.dita`. (OK it has a lot but I can reproduce with just one.)
- That topic has a `<xref>` to `othertopic.dita` or `../productB/othertopic.dita`
- I want the output to contain the rendered HTML version of `original.dita`, which pulls link text from `othertopic.dita`, but I do not want `othertopic.dita` to generate HTML
 
Results today:
- If I build the map to HTML5, my output contains the original and the other topic
- If I build the map to HTML5 with `onlytopic.in.map=true`, I don't get the other topic. BUT, the other topic isn't copied into the temp dir, so I get bad link text, and errors from the `topicpull` step about a missing file.
- If I add the topic to my map and set `processing-role="resource-only"`, regardless of the `onlytopic.in.map` option, I get a generated copy of the other topic.

**Comment by drmacro** [comment]
I don't think your use case is valid if you are using a direct URI reference on the xref.

If you are using direct URI references then I think all bets are off because there's no way for the processor to know with certainty that the topic you're referencing by direct URI reference is the same topic as referenced by a topicref and certainly no obvious requirement that processor must be able to make such a correlation.

That is, unless you are using a key reference to the resource-only topicref, there's no obvious way for a processor to know that your xref intends to be a reference to the topic as referenced by that resource-only topicref. Because the direct URI reference provides no use context or authorial intent there's no way for a processor to make any reliable assumptions about the topic as it relates the uses of the referencing topic in the current map.

For example, if the topic was a resource-only topic in one key scope but not in another, which use would you expect the processor to select, if it could select one? What if the referencing topic is not itself in either scope?

However, if the reference is to a resource-only key then resource-only status is unambiguous.

But in that case the xref cannot be resolvable so it seems to me an open question as to whether it would be appropriate or even allowed by the spec to generate text for the xref in that case (because the xref cannot be a resolvable link).

Searching on "xref" I find this text under the rules for generating text for key-based xrefs and links (that is, elements that allow  both @href and @keyref):

2.3.4.9 Processing key references to generate text or link text

...

"6. Otherwise, if the element with the key reference results in a link, normal link text determination rules apply as they would for <xref> (for example, using the <navtitle> or falling back to the URI of the link target)."

That says to me that link text is only generated if there is a link and a keyrefs to resource-only topics cannot result in a link.

But that said, I don't find any other language around xref that explicitly precludes having the xref generate text in the case where the reference is to a resource-only topic so I could buy an argument that that's an allowable thing to do.

The reference entry for xref is remarkably silent on the generated-text implications, expectations, or requirements, which surprises now that I read it. Clearly we've all fully internalized our common expectations or assumptions about xref processing.

**Comment by shaneataylor** [comment]
I think it's fair to have differing expectations about how (or if) an xref to a topic that's not part of the deliverable should be resolved. The expected OT behavior should probably be clarified, if not in the spec (as prescriptive) then in the OT documentation (as an implementation choice).

However, that's somewhat beside the point that an xref (whether direct or indirect) should not cause the target topic to be included in the deliverable if it would not otherwise be created.

**Comment by drmacro** [comment]
My point is that a direct URI reference to a topic is inherently ambiguous because there's no general or reliable way for the processor to know what the intent of the author was in terms of what use context the target topic should be considered to be in, if any. The options are:

- No context at all (the topic is not in the current map and may not be in any map at all)
- Use the first or only only normal-role use context for the topic in the same map as the referencing topic (basically what you get today if there is exactly one normal-role use of the topic)
- Use the topic as used in a different map (basically fake cross-deliverable links where you know that your processor will produce a result at the target location because that's how your processor works.
- Apply some local heuristic to try to determine the most appropriate use context for the topic based 

Nothing the in the URI of the target will tell you what the author's intent is. Therefore the best a processor can do is choose one of these options as its only behavior (or, more generally, provide configuration options ala only-topic-in-map that control how to handle this case).

In the case where the topic happens to be used exactly once in the root map then it is possible to assume that the author intended that topic and create a reliable link to it (because we know where the result will be generated because we're processing both topics as part of the same root map).

But if the topic is used two or more times then there's no way to know from the xref alone which use, if any, the author intended.

That is, what I see Robert asking for is essentially this:

For each direct URI reference to a topic from an xref, examine all topicrefs to see if they refer to this topic. If there is exactly one such reference and it is a normal-role usage then create a normal xref to this topic in this use context. If there is exactly one such reference and it is a resource-only usage then do not create a link but do set the effective text of the xref to the link text for the target. If there are multiple uses but exactly one is a normal-role usage then assume that is the use the author intended and construct the link to that use. Otherwise fail because the reference is ambiguous (there are multiple resource-only or normal-role uses, particularly multiple uses in different key scopes or filtered branches). If the topic is not used in the map at all then apply the only-topic-in-map behavior (generate a result with no map context if only-topic-in-map is "false", do not generate a result if only-topic-in-map is "true").

I can see that this behavior would produce the expected result for a common case (a topic used exactly once in a map) but will not produce a good result if the topic is used multiple times. 

The argument for this behavior is, I think "don't force people to use keys when they're not doing reuse" and I think that is the argument Robert is explicitly or implicitly making.

The argument against this behavior is "it works by accident and only under these narrow conditions and is guaranteed to fail when there is reuse, so don't ever do it.", which is my position.

But I understand my position is extreme . So if we do implement Robert's request we need to make it clear that it only works when there is no reuse and that if you are doing reuse you have to use keys or it simply can't work reliably (because the intended use context is absolutely ambiguous).

**Comment by shaneataylor** [comment]
@drmacro I'll let Robert speak for himself. My question is simple, because I am simple: is it your position that there are some circumstances when an `<xref>` alone should be sufficient to include the targeted topic in the set of generated topics copied to the output folder (i.e., the deliverable)?

**Comment by robander** [comment]
I do think your position is, as you say, extreme. And I don't think the suggested use case is particularly unreasonable. I do understand that my expectations / use cases (and those of my users) differ from yours. And while the user that came in with this particular case is not using keys, I'm not trying to make an argument for or against them.

> If you are using direct URI references then I think all bets are off because there's no way for the processor to know with certainty that the topic you're referencing by direct URI reference is the same topic as referenced by a topicref and certainly no obvious requirement that processor must be able to make such a correlation.

I just don't understand this as the basis for the argument.

- In my topic, I have `<xref href="othertopic.dita"/>`. 
- I don't care what other topic references exist to that topic, or if it is ever referenced by a topicref.
- This is a valid local reference to a file that exists. It's semantically a cross reference, the same way that a link to `otherthing.blat` would be.
- But because it's DITA, the toolkit has always been able to pull in link text - I don't view that as an accident, I view it as expected behavior.
- If that's the case, the only remaining question is what that link should go to. At this point, I don't think the use of keys (or not) is relevant:
-- If using keys: then I presumably have a `<keydef>` for it. But that's just another variant of the use case I described: the topic is referenced from the map as resource-only, that reference resolves to `othertopic.dita`, and today, I get a rendered HTML version of my resource-only topic.
-- If not using keys, it's the use case I described above - I can't find a way to 1) get the link text and 2) not get the topic
- Either way, today, I end up with a link to the topic I want (and again, I don't think this is an accident, I think it's the best default in a case where the link target was not modified by keys / scopes). It's just that I today I can't control whether that topic generates its own output in HTML.

> That is, what I see Robert asking for is essentially this:

I'm ... not sure if what followed was correct. It feels *waaaaay* more complicated than what I am expecting. True, I'm not presently concerned about keys or key scopes, but I still don't think they are relevant to this. 

Whether you use keys or not, this boils down to: how can I have a link or inline cross-reference to a local topic, where I get the link text (because the file is there!), but not generate output for that topic? I feel this is what the `onlytopic.in.map` parameter and the `processing-role="resource-only"` attribute are designed for, but don't seem to be doing anything here.

**Comment by robander** [comment]
Sort of a summary...

- Linking to a local DITA topic is legal
- That topic doesn't have to be in the map, or can be "resource-only" from the map
- The toolkit should have some response to this, whether or not you consider the scenario a best practice. [We have a lot of scenarios that use this approach; the one I described above is simply boiled down to the simplest version.]
- I don't think the current defaults make sense, particularly WRT `resource-only` -- which is a way of saying "I'm going to use this topic for something [link grabbing link text], but don't render it"

**Comment by xephon2** [comment]
#2493 is related

**Comment by drmacro** [comment]
I think the issue here devolves specifically from the unique characteristic of HTML, which is that it is meaningful (or at least possible) to render topics to HTML files outside of any map context. That is, there are clearly different expectations and different processing possibilities for multi-file outputs vs. monolithic outputs. 

I think there are really three issues here

1. Should the OT retrieve link text for xrefs that point to resource-only topics? This is independent of the form of address. 
2. How should the OT choose a target rendition of a given topic referenced by direct URI reference?
3. How should the OT behave when there is a direct URI reference to a topic that is not also used from the map? This should be independent of rendition type (that is, a solution should be capable of providing working solutions for any rendition type, multi-file or monolithic).

For item (1) it seems reasonable to me to get the link text even though the link itself either cannot be a navigable link (monolithic outputs) or ends up being a link to a rendition of the topic that does not reflect the current map (e.g., the topic rendered in isolation or the topic as rendered in the context of a different map).

For item (2) I think it's reasonable to have a rule that says "if there's exactly one navigation use of the topic, link to that rendition of the topic" since that's always going to be the right answer in this case. 

But if there are multiple uses of the topic the OT has to at least produce a warning that the reference is ambiguous and that the OT chose one particular use arbitrarily (e.g., the first use in map order since that's what the OT has always done {but it might be more useful to choose the closest use based on map structure since that's probably what the author actually wants}).

For item (3) I think there's a deeper discussion that has to happen and that the results will be different for different output types. I'm not sure I care because I consider this form authoring deeply misguided and wrong and therefore effort spent supporting it is probably not a good use of time. But I also realize that there millions of topics worth of documents that do this today so there has to be some support for it.

I also realize that my thinking about linking is expressly limited to processing in the context of maps, where you both have distinct use contexts and can both distinguish and point to them by using keys.

[I wrote all of the following before trying a simpler response, it might be useful as background so I'm leaving it here.]

My point is that you have to know the use context of the target topic in order to know, for a given source topic referenced by a direct URI reference, *which use of it you want to link to in the rendered result.*

A direct URI reference tells you nothing about the use context, therefore you *cannot know* from the reference alone which *rendered version* (or versions) of the topic the link should resolve to in the rendered version of a topic.

You can author a link to any topic using a direct URI reference *but there's no guarantee that there is a meaningful rendering or resolution of that reference* because you don't (and can't) know the use context.

In the specific case where a given topic is used exactly once in the *same map* as the referencing topic then you can reasonably make some assumptions about the use context, namely, because there is only one use context you can reasonably presume that that's the use context the author of the link expected to use.

But even that assumption is suspect because the author of the link has *no way of knowing*, in the general case, whether or not the target topic is ever used from a given map using their topic or whether it is used exactly once. 

This is the fundamental problem that DITA exposes: as the author of a topic YOU CANNOT KNOW and CANNOT PREDICT where that topic or any other topic you might link to will be used. Therefore your only reliable option as a topic author is to link to specific uses of the topic by using key references to navigation keys. 

Even if you can reliably know the use context of a target topic at the time you author the link you cannot know it for any time in the future because anyone can add a new use of the topic at any time (or remove existing uses). In a collaborative environment you can't even be sure that what you think you know at the time of authoring is accurate because somebody might be committing a change to maps you just inspected between the time you inspect the maps and the time you author your link.

If I have a direct URI xref reference to a topic that is used in two differently-filtered branches where the topic title will be different in those two branches, which filtered version of that topic should the xref reflect?

What possible general algorithm would always give the correct answer (meaning correctly reflect the author's intent in that case)?

The answer is "there cannot be such an algorithm because it is impossible to know the author's intent from the URI reference alone because the URI reference provides no information about the intended use context of the topic."

I guess it can make sense to use an xref to a resource-only topic to simply reflect the topic title even though there can't be a navigable link in the output (because the target topic is not rendered in the context of the map containing the referencing topic that also only uses the target topic as a resource).

But even then, the issues with use context ambiguity apply if there is branch filtering or key scopes.

When you say "that topic doesn't have to be in the map" I disagree: if the target topic is not in the map that contains the referencing topic (meaning isn't referenced by a navigation-role topicref) then, in the context of that map *it has no existence* and there is *no rendered form* of that topic to which a link could reliably be made. 

In the context of monolithic output like PDF it is absolutely the case that the topic will not be available in the rendered result and therefore a link to it *could not possibly work*. In the case of multi-file outputs like HTML it might be the case that some other processing instance will produce a rendered version of the topic at some predictable location but there's no way to know that, in general, from the processing of the current map and therefore, as an author, you can't rely on that behavior (and in any case other renditions would be guaranteed failure so why would you ever do that, as an author)? Or the processor could render topic in isolation, no map context, which I guess might be appropra

That is, it doesn't matter whether or not the DITA spec allows it, a direct URI reference to a topic not in a map cannot possibly work reliably in all renditions and it can only work in some renditions by accident.

That's why we added keys and why we added cross-deliverable linking: so as an author you can guarantee that your links have unambiguous targets that can work in all renditions.

**Comment by robander** [comment]
I'm going with your simpler response rather than the background :-)

My thoughts of course are influenced by my past experience / what my own users expect. But also, for me, a lot of it comes down to the "keep it simple" philosophy -- as much as we can, just do what the users expect, as long as it is reasonable / is allowed by the spec.

> 1. Should the OT retrieve link text for xrefs that point to resource-only topics? This is independent of the form of address.

Sounds like we agree on this one - it should.

>  2.  How should the OT choose a target rendition of a given topic referenced by direct URI reference?

I think we agree here as well - if there's only one rendition, link to it. If there are multiple, and we can detect that our link is ambiguous, a warning is helpful.

> 3. How should the OT behave when there is a direct URI reference to a topic that is not also used from the map? This should be independent of rendition type (that is, a solution should be capable of providing working solutions for any rendition type, multi-file or monolithic).

I don't think the answer is independent of rendition type. But I think we (more or less) agree here too, if I properly understand your own response to this question.

This view is entirely based on what my quite large user community expects (and has gotten, thanks to our heavily hacked DITA-OT processing):

- Plain X/HTML: If I have a local link to a topic that's not in the map, we pull link text and generate the link, with no warning. This is extremely common, with our history of Eclipse plug-in based delivery. Plugins link to each other, are built independently, and have stable file names. For our most common cases, this means both the author and the toolchain can reliably predict where topics will be used and how they will be named.
 This is not misguided or wrong, and we have millions of topics doing this, going back 10+ years. Obviously the older source did it with direct URIs; many projects, but definitely not all, now do the same thing with keys. 
- X/HTML formats where the result is monolithic (HTML Help, EPUB, etc) - warning is needed. 
- Same for any X/HTML format where result file names aren't known - warning is needed
- For PDF, from day 1, we've warned about links to local topics that are not part of the PDF.

Going back to our beta DITA days around 2001, I've encountered only a handful of people who were surprised by or wanted to change those defaults.

**Comment by drmacro** [comment]
So your users who are doing it (references to topics) with keys can't be referencing topics not in the map, because before DITA 1.3 there was no way to do that. At worst the reference is to a resource-only topic and in that case there can be no link (because the target is explicitly known to be resource-only and should never, for that use, have any direct reflection in any result generated from that map).

As long as the warnings you outlined are provided then I'm OK with the proposed solution. My main concern has always been that the system will make it look like direct-URI references *should* work when they are ambiguous. 

But note that this still leaves open the general problem of how to know, for a given direct URI reference to a topic, what topicrefs, if any, also refer to that topic. 

A general solution has to work in the face of non-file-system resolution of URIs (meaning you don't have a file object you can compare, you only have the URI itself or the result returned by the URI resolver, which will only be a byte sequence).

So you have to compare URIs which means you have to normalize them in some way. Not sure there's a general solution for that.

If you want to stipulate that this only works reliably when operating on files in a file system or where the storage system provides some kind of reliable object ID as part of the URI I'm OK with that. I just want us all to be clear that general solution requires the ability to reliably compare URIs to determine that a given direct-URI reference is to a topic referenced from topicrefs in the map.

**Comment by robander** [comment]
I sort of fear this has gotten off track because you didn't like my use case. The original report above is pretty straightforward, and is still (I think) the most important thing we can fix.

If I have this in my map, as my only reference to `just-a-resource.dita`:
`<topicref href="just-a-resource.dita" processing-role="resource-only"/>`

That topic reference, when built to X/HTML, results in a generated file. It should not, because the document is marked resource-only.

I don't think that any other stipulations / caveats are necessary.

**Comment by drmacro** [comment]
I definitely agree with that: a resource-only topicref to a topic should never result in a resulting HTML file (or any direct reflection in any result of any type).

**Comment by taggar** [comment]
We're experiencing the same issues: resource-only files are ending up as files in the html output, plus we also have tons of empty and unused "Chunk<number>.htm* files. Quite annoying since it makes for a lot of "noise" plus the resource files also end up being crawled by the search engine and are being presented to end-users as hits when they shouldn't be ...

**Comment by robander** [comment]
Added sample files to the description. Samples include:

- `2077.ditamap` as the root map
- A normal topic
- A resource-only topic
- A topic that is linked to by `normal.dita` but is not in the map; another that is linked-to and in the map as `resource-only`
- A topic that is linked to by `resource-only.dita` but is not in the map; another that is linked-to and in the map as `resource-only`
- A topic that is a conref target for `normal.dita` but is not in the map; another that is a conref target and in the map as `resource-only`
- A topic that is a conref target for `resource-only.dita` but is not in the map; another that is a conref target and in the map as `resource-only`

Currently, conref targets and link targets that are _not_ listed in the map generate output. I don't like this behavior but I guess it is expected based on past behavior. (I don't think the default should render conref targets that are not in the map, but that's a discussion for elsewhere.)

Link targets from both the normal and from the resource-only topic are currently rendered, whether or not they are marked as `resource-only` in the map. This is what causes most of our problems. I have a topic that is only a conref target -- has lots of reusable pieces. Somewhere in that topic, I have links. Currently, I cannot avoid generating the target of those links -- even if I list them as resource-only, and even if I _never even generate those links_. 

Example: my conref library has a link somewhere to `random.dita`. I list the conref library as `resource-only` so it's not generated. I only ever use a product name from that library. Today in my HTML build, I cannot avoid generating `random.html`, even though no rendered topic links to it, even if I mark that same topic as `resource-only`.

**Comment by robander** [comment]
Closed with #2694
